plugins/radar-android-phone/src/main/java/org/radarbase/passive/phone/PhoneBluetoothManager.kt

```diff
@@ -79,20 +105,59 @@ class PhoneBluetoothManager(service: PhoneBluetoothService) : AbstractSourceMana
             bluetoothBroadcastReceiver = object : BroadcastReceiver() {
                 private var numberOfDevices: Int = 0
 
+                val hasConnectPermission = Build.VERSION.SDK_INT < Build.VERSION_CODES.S
+                        || ActivityCompat.checkSelfPermission(service, Manifest.permission.BLUETOOTH_CONNECT) == PackageManager.PERMISSION_GRANTED
+
+                @SuppressLint("MissingPermission")
                 override fun onReceive(context: Context, intent: Intent) {
                     val action = intent.action ?: return
                     when (action) {
                         BluetoothDevice.ACTION_FOUND -> {
                             numberOfDevices++
+
+                            val device: BluetoothDevice = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                                intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE, BluetoothDevice::class.java)
+                            } else {
+                                @Suppress("DEPRECATION")
+                                intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
+                            } ?: return
+
+                            val macAddress = device.address
+                            val macAddressHash: ByteBuffer = hashGenerator.createHashByteBuffer(macAddress + "$hashSaltReference")
+
+                            val scannedTopicBuilder = PhoneBluetoothDeviceScanned.newBuilder().apply {
+                                time = currentTime
+                                timeReceived = currentTime
+                            }
+
+                            val pairedDevices: Set<BluetoothDevice> = if (hasConnectPermission) bluetoothAdapter.bondedDevices else emptySet()
+                            val newPairedDevices = pairedDevices.subtract(allPairedDevices)
+                            allPairedDevices.addAll(newPairedDevices)
+                            allPairedDevices.removeAll(allPairedDevices.subtract(pairedDevices)) // removing devices which are previously paired but unpaired now
```

Comment:
isn't this essentially just keeping the current pairedDevices in allPairedDevices? So why do we do all the subtract/add ?

Reply:
why not just keep the latest list in that case? `allPairedDevices = pairedDevices`